(Network) Implement binary network streams

Cthulhu-throwaway opened the pull request:
## Description

Network streams make it a lot easier in ensuring correct endianness.

Example usage:
```c
uint8_t     buf[1024];
uint8_t     random[256];
uint32_t    magic;
void       *data;
size_t      len;
netstream_t stream;

netstream_open(&stream, buf, sizeof(buf), 0);
netstream_write_dword(&stream, 'PING');
netstream_write(&stream, random, sizeof(random));
netstream_data(&stream, &data, &len);
socket_send_all_blocking(fd, data, len, true);
netstream_close(&stream, false);

len = sizeof(magic) + sizeof(random);

socket_receive_all_blocking(fd, buf, len);
netstream_open(&stream, buf, sizeof(buf), len);
netstream_truncate(&stream, len - 8);
netstream_read_dword(&stream, &magic);
netstream_seek(&stream, -4, NETSTREAM_SEEK_END);
netstream_read(&stream, random, 0);
netstream_close(&stream, false);

if (magic == 'PONG')
   RARCH_LOG("PONG received!!!");
```

LibretroAdmin commented:
Couple of design considerations:

* All the other stream-based files are in streams/. Not sure if we should put this one in network/ instead. 
* file_stream and memory_stream mostly have the same kind of interface so that they can be used as VFS backends. Do you think it would be possible to do this with network_stream too? It'd require interface_stream to be extended to add support for the network backend. 

Totally up to you though. These are just ideas off the top of my head.

Cthulhu-throwaway commented:
> * All the other stream-based files are in streams/. Not sure if we should put this one in network/ instead.

I can agree with that.
Moved to libretro-common/streams, moved the source code into its .c file (getting rid of the inline attributes) and worded the function names to better resemble those of other streams.

> * file_stream and memory_stream mostly have the same kind of interface so that they can be used as VFS backends. Do you think it would be possible to do this with network_stream too? It'd require interface_stream to be extended to add support for the network backend.

Took a look at the stream interface and I didn't like what I saw.
Pointing a few:

- Most of the code in there is not suitable for a network stream.
- The stream interface can only work with dynamic memory. Taking netplay as an example, most of the time we can use a stack buffer to send and receive data, thus bypassing the need for slow memory allocations and deallocations, aswell as memory fragmentation associated with dynamic memory buffers.
- Lots of work needed to extend it to support what we need in a network stream and in the end, not much will be gained in the process.
- Lots of TODO's.

Overall, while a network stream is just an abstraction of a memory stream, they work differently; network streams must always swap the bytes of multi-byte types (if the machine is little-endian) and there are certain tasks that we could perform on a normal memory stream but are too unsafe to be performed on data that just arrived on the network.
Implementing it directly into the current stream interface would mean nop'ing the majority of the interface's function (stuff like getc, putc, etc doesn't make sense for a network stream) and implementing functions that would only be used by the network stream.
Unless there is/are highly convincing reason(s) for me to implement this directly into the stream interface, I feel like its current ad hoc serves its purpose correctly.

LibretroAdmin commented:
OK, fair enough.
